File: cpp98/pi-exchange/pi_sender.cpp
/*

Copyright © 2017-2018,
Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC
All rights reserved. See LICENSE file and DISCLAIMER for more details.
*/
static char help[] = " PI SENDER: Simple program to demonstrate the usage of HELICS C Interface.\n\
            This example creates a ZMQ broker and a value federate.\n\
            The value federate creates a global publications and publishes\n\
            t*pi for 20 time-steps with a time-step of 0.01 seconds.\n\n";


#include <stdio.h>
#include <helics/cpp98/ValueFederate.hpp>
#include <helics/cpp98/Broker.hpp>
#include <helics/cpp98/helics.hpp> // helicsVersionString

int main(int /*argc*/,char ** /*argv*/)
{
  std::string    initstring="-f 2 --name=mainbroker";
  std::string    fedinitstring="--federates=1";
  double         deltat=0.01;
  helicscpp::Publication pub;

  std::string helicsversion = helicscpp::getHelicsVersionString();

  printf("PI SENDER: Helics version = %s\n",helicsversion.c_str());
  printf("%s",help);

  /* Create broker */
  helicscpp::Broker broker("zmq","",initstring);

  if(broker.isConnected()) {
    printf("PI SENDER: Broker created and connected\n");
  }

   /* Create Federate Info object that describes the federate properties
    * Sets the federate name and core type from string
    */
  helicscpp::FederateInfo fi( "zmq");

  /* Federate init string */
  fi.setCoreInit(fedinitstring);

  /* Set the message interval (timedelta) for federate. Note that
     HELICS minimum message time interval is 1 ns and by default
     it uses a time delta of 1 second. What is provided to the
     setTimedelta routine is a multiplier for the default timedelta.
  */
  /* Set one second message interval */
  fi.setProperty(helics_property_time_delta, deltat);
  fi.setProperty(helics_property_int_log_level, helics_log_level_warning);

  /* Create value federate */
  helicscpp::ValueFederate* vfed = new helicscpp::ValueFederate("Test sender Federate", fi);
  printf("PI SENDER: Value federate created\n");

  /* Register the publication */
  pub = vfed->registerGlobalPublication("testA", helics_data_type_double);
  printf("PI SENDER: Publication registered\n");

  /* Enter initialization state */
  vfed->enterInitializingMode(); // can throw helicscpp::InvalidStateTransition exception
  printf("PI SENDER: Entered initialization state\n");

  /* Enter execution state */
  vfed->enterExecutingMode(); // can throw helicscpp::InvalidStateTransition exception
  printf("PI SENDER: Entered execution state\n");

  /* This federate will be publishing deltat*pi for numsteps steps */
  //double this_time = 0.0;
  double value = 22.0/7.0;
  helics_time currenttime=0.0;
  int           numsteps=20,i;

  for(i=0; i < numsteps; i++) {
    double val = currenttime*value;

    printf("PI SENDER: Sending value %3.2fpi = %4.3f at time %3.2f to PI RECEIVER\n",deltat*i,val,currenttime);
    pub.publish( val);

    currenttime = vfed->requestTime(currenttime);
  }

  vfed->finalize();
  printf("PI SENDER: Federate finalized\n");

  // destructor must be called before call to helicsCloseLibrary(), or segfault
  delete vfed;

  broker.waitForDisconnect();

  printf("PI SENDER: Broker disconnected\n");
  helicsCloseLibrary();
  return(0);
}

File: python/delayed-pi-exchange-filter/pisender.py
# -*- coding: utf-8 -*-
import time
import helics as h
from math import pi
import pprint

initstring = "-f 3 --name=mainbroker"
fedinitstring = "--broker=mainbroker --federates=1"
deltat = 0.01
# Setting up pretty printing, mostly for debugging.
pp = pprint.PrettyPrinter(indent=2)



def eval_data_flow_graph(fed):
    query = h.helicsCreateQuery("broker", "data_flow_graph")
    graph = h.helicsQueryExecute(query, fed)
    print(f'PI SENDER: Data flow graph:\n {pp.pformat(graph)}')
    return graph



def eval_dependency_graph(fed):
    query = h.helicsCreateQuery("broker", "dependencies")
    graph = h.helicsQueryExecute(query, fed)
    print(f'PI SENDER: Dependency graph: \n {pp.pformat(graph)}')




helicsversion = h.helicsGetVersion()

print("PI SENDER: Helics version = {}".format(helicsversion))

# Create broker #
print("Creating Broker")
broker = h.helicsCreateBroker("zmq", "", initstring)
print("Created Broker")

print("Checking if Broker is connected")
isconnected = h.helicsBrokerIsConnected(broker)
print("Checked if Broker is connected")

if isconnected == 1:
    print("Broker created and connected")

# Create Federate Info object that describes the federate properties #
fedinfo = h.helicsCreateFederateInfo()

# Set Federate name #
h.helicsFederateInfoSetCoreName(fedinfo, "pisender")

# Set core type from string #
h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

# Federate init string #
h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

# Set the message interval (timedelta) for federate. Note th#
# HELICS minimum message time interval is 1 ns and by default
# it uses a time delta of 1 second. What is provided to the
# setTimedelta routine is a multiplier for the default timedelta.

# Set one second message interval #
h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, deltat)

# Create value federate #
vfed = h.helicsCreateCombinationFederate("pisender", fedinfo)
print("PI SENDER: Combo federate created")

epid = h.helicsFederateRegisterGlobalEndpoint(vfed, "pisender_ep", "")
print("PI SENDER: Endpoint registered")

# Evaluating data-flow and dependency graphs 
# Manually pausing to ensure all other federates have configured as well
time.sleep(5)
eval_data_flow_graph(vfed)
eval_dependency_graph(vfed)

# Enter execution mode #
h.helicsFederateEnterExecutingMode(vfed)
print("PI SENDER: Entering execution mode")

# This federate will be publishing deltat*pi for numsteps steps #
this_time = 0.0
value = pi

for t in range(5, 10):
    val = value

    currenttime = h.helicsFederateRequestTime(vfed, t)

    #h.helicsPublicationPublishDouble(pub, val)
    print(
        "PI SENDER: Sending value pi = {} at time {} to PI RECEIVER".format(
            val, currenttime
        )
    )

    h.helicsEndpointSendEventRaw(epid, "pireceiver_ep", str(val), t)
    time.sleep(1)

h.helicsFederateFinalize(vfed)
print("PI SENDER: Federate finalized")

while h.helicsBrokerIsConnected(broker):
    time.sleep(1)

h.helicsFederateFree(vfed)
h.helicsCloseLibrary()

print("PI SENDER: Broker disconnected")
